sys/net/link_layer/ble_ll/ble_ll.c

```diff
@@ -0,0 +1,331 @@
+         void *value, size_t max_len)
+{
+    switch (opt) {
+        case NG_NETOPT_ADV_ADDRESS:
+            return _get_adv_address(value, max_len);
+
+        default:
+            return dev->driver->get(dev, opt, value, max_len);
+    }
+}
+
+/**
+ * @brief   BLE link layer send function
+ */
+static int _send(ng_netdev_t *dev, ng_pktsnip_t *pkt)
+{
+    uint8_t payload[BLE_MAX_PKT_LEN];
+    ng_pktsnip_t *ble_pkt = NULL;
+    uint8_t ble_pkt_size;
+
+    if (pkt == NULL) {
+        DEBUG("ble_ll: Error handling packet: packet incomplete\n");
+        return -ENOMSG;
+    }
+
+    ble_pkt_size = ng_pkt_len(pkt);
+
+    if (ble_pkt_size > BLE_MAX_PKT_LEN) {
+        ng_pktbuf_release(pkt);
+        DEBUG("ble_ll: Error sending packet: invalid BLE pdu header length\n");
+        return -EOVERFLOW;
+    }
+
+    memcpy(payload, pkt->data, BLE_PDU_HDR_LEN);
+    /* write advertising address to payload */
+    memcpy((payload + BLE_PDU_HDR_LEN), _adv_addr, BLE_ADDR_LEN);
+    ble_pkt_size += BLE_ADDR_LEN;
+
+    /* write optional payload data */
+    if (pkt->next != NULL) {
+        ble_pkt_size += ng_pkt_len(pkt->next);
+        if (ble_pkt_size > BLE_MAX_PKT_LEN) {
+            ng_pktbuf_release(pkt);
+            DEBUG("ble_ll: Error handling packet: payload too large\n");
+            return -EOVERFLOW;
+        }
+        memcpy((payload + BLE_PDU_HDR_LEN + BLE_ADDR_LEN),
+                pkt->next->data, ble_pkt_size);
+    }
+
+    ble_pkt = ng_pktbuf_add(NULL, payload, ble_pkt_size, NG_NETTYPE_UNDEF);
+    ng_pktbuf_release(pkt);
+
+    return dev->driver->send_data(dev, ble_pkt);
+}
+
+/**
+ * @brief    Handle incoming link layer packet
+ */
+static void _receive(ng_pktsnip_t *pkt)
+{
+    ble_pkt_t *data;
+
+    data = (ble_pkt_t *)pkt->data;
+
+    if (_ll_state == LL_ADVERTISING) {
+        /* Send respone to received BLE SCAN_REQ packet */
+        if (data->header.pdu_type == SCAN_REQ_TYPE) {
+            DEBUG("ble_ll: receive_data -> SCAN_REQ received\n");
+            /*_send_scanresponse_pkt();*/
+        }
+
+        /* Send respone to received BLE CONNECT_REQ packet */
+        else if (data->header.pdu_type == CONNECT_REQ_TYPE) {
+            DEBUG("ble_ll: receive_data -> CONNECT_REQ received\n");
+        }
+    }
+
+#if ENABLE_BLE_PKT_DEBUG
+    int i;
+    static const char *pdu_type_name[] = {
+        "ADV_IND", "ADV_DIRECT_IND", "ADV_NONCONN_IND", "SCAN_REQ", "SCAN_RSP",
+        "CONNECT_REQ", "ADV_SCAN_IND"
+    };
+
+    printf("Type: %s  Len: %u %s Address: %02x:%02x:%02x:%02x:%02x:%02x\n",
+           pdu_type_name[data->header.pdu_type], data->header.length,
+           (data->header.pdu_type == SCAN_REQ_TYPE) ? "Scanning" : "Advertising",
+           data->payload[5], data->payload[4], data->payload[3], data->payload[2], data->payload[1],
+           data->payload[0]);
+
+    for (i = 0; i < data->header.length; i++) {
+        printf("%02x ", data->payload[i]);
+    }
+
+    printf("\n");
+#endif
+
+}
```

Comment:
@OlegHahm do you have a clue how to realize that logging in a "RIOT-way"? I'm just asking after all past discussions about the logging API in `core` or `DEBUG` in gnrc..